src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    internal class SequencePipeReader : PipeReader
+    {
+        private ReadOnlySequence<byte> _sequence;
+        private bool _isReaderCompleted;
+
+        private int _cancelNext;
+
+        public SequencePipeReader(ReadOnlySequence<byte> sequence)
+        {
+            _sequence = sequence;
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            ThrowIfCompleted();
+
+            // Fast path: did we consume everything?
+            if (consumed.Equals(_sequence.End))
+            {
+                _sequence = ReadOnlySequence<byte>.Empty;
+                return;
+            }
+
+            _sequence = _sequence.Slice(consumed);
+        }
+
+        /// <inheritdoc />
+        public override void CancelPendingRead()
+        {
+            Interlocked.Exchange(ref _cancelNext, 1);
```

Comment:
@davidfowl, I'm curious, why are the semantics of Cancel**Pending**Read that it cancels the next read if there isn't one pending?

Reply:
The next read will be cancelled.

Reply:
I understand.  I'm asking why it was designed that way :)

Reply:
CancelPendingRead is often used for waking up the loop in order to do something and this makes it easier to write code that doesn't need to do any other synchronization in order to make sure that the cancellation will be observed.

Reply:
One concrete example of this is Kestrel's graceful shutdown. [We unset a `_keepAlive` flag and call CancelPendingRead](https://github.com/dotnet/aspnetcore/blob/a104f9d3c7990446326ca03d1dd777f41a6c4f53/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs#L115-L124) on every connection's PipeReader. This guarantees that every connection observes the change to the `_keepAlive` flag even if the connection has not yet started reading the next request.

Reply:
That use makes sense.  I'd just have expected the most desirable semantic to be either a) cancel only the currently pending read or b) cancel the currently pending read and all future reads, but it's actually c) cancel the currently pending read or just the next read if one isn't currently pending.  I'd have though the graceful shutdown use case called out would have been better with (b), but in practice it probably doesn't matter.  I'd also have expected a different name than CancelPendingRead :smile:

Reply:
The reason we don't want it to cancel all future reads is in case we're in the middle of reading a request. In that case, we observe the cancellation but essentially ignore it until we're done with the request.

Reply:
Thanks.  I spoke with @halter73 offline and he explained this to me in more detail.  I understand the intent now.